Bluetooth: SMP: Handle errors during pairing and encryption

joerchan opened the pull request:
Give pairing failed reason in the pairing_failed callback.
Provide pairing_failed and pairing_complete callback for SSP
Fixes: #17013

Stop pairing procedure in the request phase, instead of the key distribution phase. This solves one instance where we would end up with pairing failed after the link has been encrypted.
Provide a callback for encrypting the link has failed. This gives the application feedback when a bond exists (or assumed to exists) for the connected peer, but the peer does not have the bond information anymore.
Fixes: #17920 

Treat SMP debug keys normally, in order to be able to debug a bond properly even when debug keys are used.

joerchan committed:
Bluetooth: SMP: Add pairing failed reason

Forward the pairing failed SMP status code to the application

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Vudentz commented on samples/bluetooth/peripheral_sc_only/src/main.c:
```diff
@@ -102,9 +102,9 @@ static void pairing_complete(struct bt_conn *conn, bool bonded)
+static void pairing_failed(struct bt_conn *conn, u8_t reason)
 {
-	printk("Pairing Failed. Disconnecting.\n");
+	printk("Pairing Failed (%#02x). Disconnecting.\n", reason);
```
Id include a "0x" when printing the reason to indicate it is a hex value.

joerchan replied:
That is what `#` does :)

jhedberg replied:
*mind blown* - I didn't realise there was a standard for this - I even had to go double check that the various libcs I use support this, and they do - minimal libc included. For consistency with the existing code base 0x%02x would be the right choice, but I think it does make sense to convert gradually the code base to use # and we have to start somewhere.

jhedberg replied:
summarising discussion from slack: we'll stick to the old style since using # produces inconsistent results due to exceptional handling of the value 0

Vudentz replied:
Have never seem that used before, yet it is documented in the man page of printf:

```
       #      The value should be converted to an "alternate form".  For o conversions, the first character of the output string is made zero (by  prefixing  a  0  if  it  was  not  zero
              already).   For  x  and  X  conversions,  a nonzero result has the string "0x" (or "0X" for X conversions) prepended to it.  For a, A, e, E, f, F, g, and G conversions, the
              result will always contain a decimal point, even if no digits follow it (normally, a decimal point appears in the results of those conversions only  if  a  digit  follows).
              For g and G conversions, trailing zeros are not removed from the result as they would otherwise be.  For other conversions, the result is undefined.
```
So yeah, we have been using 0x for ages without knowing # would do a better job converting the format.

jhedberg replied:
@Vudentz please check the backlog on #bluetooth on slack - we came to the conclusion that even though # is a standard it'll produce inconsistent results, i.e. we'll stick to explicitly specifying 0x for now

jhedberg reviewed (changes requested):
@joerchan this callback is not SMP specific so I don't think it's right to pass it SMP status codes. The reason you don't see the issue is that it seems we have a bug in the BR/EDR event handlers in hci_core.c, such as `ssp_complete()` in that they don't call the pairing_failed even though they should.

I'm not sure if you saw the original discussion that led to the creation of the GitHub issue, but my proposal was to create a new enum for this which you would then map to from SMP, BR/EDR (HCI) or any other errors that may occur (say, the link gets disconnected in the middle of an SMP session - what SMP code would you choose?)

joerchan commented:
@jhedberg Ok. I didn't see the original discussion. So I didn't understand why the issue mentioned an enum. I just thought it meant that the SMP error codes needed to be available to the application.
So the issue is that there could possibly be HCI error codes because of SSP (which I know nothing about).
In that case we would have to know the possible HCI error codes that could be received.
Either we map those to SMP error codes (SMP_ERR_UNSPECIFIED for those that cannot be mapped), or we combine SMP error codes and HCI authentication related error codes into an enum.
In either case we need to know which error codes we need to handle in SSP complete HCI event. Do you have any ideas which those are?

joerchan committed:
Bluetooth: host: Pairing callback fail and complete for SSP

Proved the Authentication callback for pairing failed and pairing
complete when BR/EDR SSP is complete.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented on include/bluetooth/smp.h:
```diff
@@ -0,0 +1,42 @@
+	BT_SSP_ERR_PAIRING_NOT_ALLOWED =        0x83,
+	BT_SSP_ERR_PIN_OR_KEY_MISSING =         0x84,
+	BT_SSP_ERR_UNSPECIFIED =                0x85,
+	BT_SSP_ERR_INVALID_PARAMS =             0x86,
```
What I had in mind is that we'd have a higher level BT_PAIR_ERR_* enum that the protocol (SSP, SMP) specific error would be mapped to. I was thinking we wouldn't have to do a 1:1 mapping for everything from the start, rather do the ones that we need specific error for and use some BT_PAIR_ERR_UNSPECIFIED for the rest. However, seeing that the list isn't actually that long, perhaps we can just introduce dedicated errors from the start. Any reasons why you think this approach wouldn't be a good idea?

joerchan replied:
I'm not a big fan of mapping Error codes to new error codes. In my experience it makes it harder to figure out the root cause when you have to go through multiple levels of error codes. And error codes are mainly used for debugging. Which is why I think this is best, the SMP error codes are kept. And the ones for SSP is mapped by name to the HCI names

Also (and I might be wrong) but it seems that SSP is the old way to do it. And that SMP should be used for newer version of BREDR.

If you still think creating BT_PAIR_ERR_* is better let me know, and I can switch to that, but I prefer this way.

jhedberg commented:
> Also (and I might be wrong) but it seems that SSP is the old way to do it. And that SMP should be used for newer version of BREDR.

@joerchan that's not the case AFAIK. SMP is used over BR/EDR for cross-transport key generation (Link Key -> LTK) but not for the actual pairing.

As for the the rest, feeling a bit conflicted about this - I understand your reasoning, but having a single enum where some of the values are translations and some aren't doesn't feel very clean. I'd appreciate any thoughts from @sjanc on this since he's behind most of the existing SMP code.

joerchan commented:
@jhedberg I understand, I don't know how BREDR works, and I cannot find any good MSC that describes the likely scenarios. So from side the SSP errors are purely guesswork, I'm not sure I got all the relevant error codes.

How about mapping it like this:
```
BT_PAIR_ERR_UNSPECIFIED
	HCI: Default
	SMP: Default
BT_PAIR_ERR_AUTHENTICATION_FAIL
	HCI: AUTHENTICATION_FAIL
	SMP: PASSKEY_ENTRY_FAILED
	     CONFIRM_FAILED
	     NUMERIC_COMP_FAILED
	     DHKEY_CHECK_FAIL
BT_PAIR_ERR_AUTHENTICATION_MISSING
	HCI: PIN_OR_KEY_MISSING
	SMP: OOB_NOT_AVAILABLE
BT_PAIR_ERR_AUTHENTICATION_REQUIREMENT
	HCI: None
	SMP: AUTH_REQUIREMENTS
             ENC_KEY_SIZE
BT_PAIR_ERR_NOT_SUPPORTED
	HCI: PAIRING_NOT_SUPPORTED
	SMP: CMD_NOT_SUPPORTED
BT_PAIR_ERR_NOT_ALLOWED
	HCI: NOT_ALLOWED
	SMP: BREDR_PAIRING_IN_PROGRESS
	     CROSS_TRANSPORT_NOT_ALLOWED
	     REPEATED_ATTEMPTS
BT_PAIR_ERR_INVALID_PARAMS
	HCI: INVALID_PARAMS
	SMP: INVALID_PARAMS
```

joerchan commented:
Regarding namespace, would it make more sense to have "BT_CONN_AUTH_ERR_*" and move the error codes into conn.h?

sjanc commented:
I'm fine with proposed error namespace

jhedberg commented:
> Regarding namespace, would it make more sense to have "BT_CONN_AUTH_ERR_*" and move the error codes into conn.h?

Yes, I think that makes sense as soon as it's no-longer SMP specific.

jhedberg commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1069,13 +1069,14 @@ static void auth_pairing_complete(struct bt_conn *conn, bool bonded)
 
-	shell_print(ctx_shell, "Pairing failed with %s", addr);
+	shell_print(ctx_shell, "Pairing failed with %s reason 0x%02x", addr,
+		reason);
```
Is the alignment of this split line correct? At least my browser renders it like `reason` doesn't match the ( on the previous line.

joerchan replied:
Was not aligned, fixed.

jhedberg reviewed (dismissed):
Looks ok, except for one minor coding style thing

Vudentz commented on include/bluetooth/conn.h:
```diff
@@ -297,6 +297,29 @@ int bt_conn_security(struct bt_conn *conn, bt_security_t sec);
 
+enum {
+	/* Authentication failed. */
+	BT_CONN_AUTH_ERR_AUTHENTICATION_FAIL,
```
I would omit the term AUTHENTICATION from these since it is already part of the prefix BT_CONN_AUTH.

joerchan replied:
The prefix is unfortunate in that case. This error code is to say that the authentication part of the pairing procedure failed. For example the user has compared the two different passkey and they did not match, so he pressed cancel. I feel like authentication failure is more precise than saying "fail"

Vudentz replied:
Not following you argument, the error domain is BT_CONN_AUTH_ERR and you say that is the authentication part thus I don't see why we need to keep authentication as error when the error domain is already refering to authentication part, actually even the conn part sounds a bit redundant here since authentication error can only happen on a connection.

Vudentz commented on include/bluetooth/conn.h:
```diff
@@ -297,6 +297,29 @@ int bt_conn_security(struct bt_conn *conn, bt_security_t sec);
+	BT_CONN_AUTH_ERR_AUTHENTICATION_FAIL,
+
+	/* Required input was not available.  */
+	BT_CONN_AUTH_ERR_AUTHENTICATION_MISSING,
```
Id rename this BT_CONN_AUTH_ERR_KEY_MISSING or perhaps BT_CONN_AUTH_ERR_INVALID, however I recall that some stacks do attempt to generate new keys when such error is generated, and in case of success replace the keys.

joerchan replied:
In the LE case, this would be returned when OOB data is not available. Maybe KEY_OR_DATA_MISSING?

Vudentz replied:
Im usually lean towards sorter names whenever possible as it usually produces code with least of lines and in particular less multi-line identation to deal with.

joerchan commented:
To summarize the discussion on slack.

Change the BT_CONN_AUTH_ERR_* to BT_SECURITY_ERR_*.
Instead of providing error reason in pairing_failed callback add another callback security_failed to bt_conn_cb instead. This one will be called after pairing_failed has been called.
This would be because the bt_conn_cb callbacks can be registered multiple times, and it makes more sense for the application to use connection events.
In the case where pairing_failed causes a pairing to be restarted, the pairing_failed callback is not called.

@Vudentz @jhedberg @sjanc 
I'd like to agree that this is what we want to do before I just replace the contents of this PR, I don't want to go back and forth on this.

Vudentz commented:
> To summarize the discussion on slack.
> 
> Change the BT_CONN_AUTH_ERR_* to BT_SECURITY_ERR_*.
> Instead of providing error reason in pairing_failed callback add another callback security_failed to bt_conn_cb instead. This one will be called after pairing_failed has been called.
> This would be because the bt_conn_cb callbacks can be registered multiple times, and it makes more sense for the application to use connection events.
> In the case where pairing_failed causes a pairing to be restarted, the pairing_failed callback is not called.
> 
> @Vudentz @jhedberg @sjanc
> I'd like to agree that this is what we want to do before I just replace the contents of this PR, I don't want to go back and forth on this.

We might not heard anything back from @jhedberg since he is on vacation, but perhaps @sjanc can comment on. Im fine with this direction.

carlescufi commented:
Sorry, approved by mistake.

@sjanc @jhedberg could you take a look at the [proposed solution](https://github.com/zephyrproject-rtos/zephyr/pull/17306#issuecomment-511817499) ?

carlescufi commented:
> To summarize the discussion on slack.
> 
> Change the BT_CONN_AUTH_ERR_* to BT_SECURITY_ERR_*.
> Instead of providing error reason in pairing_failed callback add another callback security_failed to bt_conn_cb instead. This one will be called after pairing_failed has been called.
> This would be because the bt_conn_cb callbacks can be registered multiple times, and it makes more sense for the application to use connection events.

@joerchan @Vudentz 
So let me get this straight. We add a new callback, `security_failed` instead of using the existing one, `pairing_failed`. This makes some sense, but beyond the naming, what is the real reason behind this? I don't quite understand the argument of the multiple registrations, and not sure why we would not have something like `security_update` callback instead?

> In the case where pairing_failed causes a pairing to be restarted, the pairing_failed callback is not called.

I assume you meant to write `the security_failed callback is not called`?

carlescufi reviewed (changes requested):
Blocking this until further clarification is provided

Vudentz commented:
> > To summarize the discussion on slack.
> > Change the BT_CONN_AUTH_ERR_* to BT_SECURITY_ERR_*.
> > Instead of providing error reason in pairing_failed callback add another callback security_failed to bt_conn_cb instead. This one will be called after pairing_failed has been called.
> > This would be because the bt_conn_cb callbacks can be registered multiple times, and it makes more sense for the application to use connection events.
> 
> @joerchan @Vudentz
> So let me get this straight. We add a new callback, `security_failed` instead of using the existing one, `pairing_failed`. This makes some sense, but beyond the naming, what is the real reason behind this? I don't quite understand the argument of the multiple registrations, and not sure why we would not have something like `security_update` callback instead?

auth_cb and conn_cb have different users, the former is single agent that deals with pairing/bonding requests while the later is what application would normally use to track connection state and security updates.
 
> > In the case where pairing_failed causes a pairing to be restarted, the pairing_failed callback is not called.
> 
> I assume you meant to write `the security_failed callback is not called`?

That is correct, the pairing agent may restart the pairing so the error don't have to be notified to security_failed, this sometimes can happen with errors such as missing link key which may lead to a new link key being generated and only of that fails security_failed would be called.

pdunaj commented:
Hi @carlescufi ,

> auth_cb and conn_cb have different users, the former is single agent that deals with pairing/bonding requests while the later is what application would normally use to track connection state and security updates.
@Vudentz  stated what we need. At the moment we don't have hooks for auth_cb. Application relies on the bt_conn_cb to check the connection status. One of the callbacks is `security_changed`. What we observed is that it was called with level indicating security being set even-though security failed. The end effect was application assuming connection is secured and starting HID data transmission to non-secured link.

We would not need new callback if the current one (`security_changed`) would inform us about the real level of security (is it passing the desired/set level now?).

Vudentz commented:
> Hi @carlescufi ,
> 
> > auth_cb and conn_cb have different users, the former is single agent that deals with pairing/bonding requests while the later is what application would normally use to track connection state and security updates.
> > @Vudentz  stated what we need. At the moment we don't have hooks for auth_cb. Application relies on the bt_conn_cb to check the connection status. One of the callbacks is `security_changed`. What we observed is that it was called with level indicating security being set even-though security failed. The end effect was application assuming connection is secured and starting HID data transmission to non-secured link.
> 
> We would not need new callback if the current one (`security_changed`) would inform us about the real level of security (is it passing the desired/set level now?).

You mean that security_changed would be called with an error if the security could not be changed? While it sounds possible Id say having a security_failed is a bit cleaner on the interface and the application may choose not to register anything if it doesn't want to as oppose to security_changed that would double as changed or failed logic.

joerchan commented:
> We would not need new callback if the current one (security_changed) would inform us about the real level of security (is it passing the desired/set level now?).

@pdunaj The `security_changed` callback has the security level, but it seems it is not always called when pairing failed. So then something needs to be done anyway.

joerchan committed:
Bluetooth: SMP: Add option to treat debug keys normally during debugging

Allow SMP debug keys to behave in the same way as normal keys, in order
to debug with encryption and Bluetooth sniffer the exact way it behaves
when not using debug keys.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented:
> @pdunaj The security_changed callback has the security level, but it seems it is not always called when pairing failed. So then something needs to be done anyway.

When pairing fails the security level doesn't change, which is why the callback is not called.

joerchan commented:
> > @pdunaj The security_changed callback has the security level, but it seems it is not always called when pairing failed. So then something needs to be done anyway.
> 
> When pairing fails the security level doesn't change, which is why the callback is not called.

No, it seems that is not always correct. See here: https://github.com/zephyrproject-rtos/zephyr/issues/17920#issuecomment-521143001

jhedberg commented:
> @jhedberg No, it seems that is not always correct. See here: #17920 (comment)

@joerchan right that's a actual bug then, however the purpose of the callback is to only be called on actual level changes.

joerchan committed:
Bluetooth: SMP: Stop new pairing early if MAX_PAIR has been reached

Stop the pairing procedure in the request phase if no storage is
available for the keys. This avoids the pairing procedure from failing
during the key distribution phase.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: Host: Add error to security changed callback

Add security error to security_changed callback. Call this callback when
security has failed and provide current security level and error.
Reason for failure can be.
 - Pairing procedure failed, pairing aborted before link encryption.
 - Link encrypt procedure failed
 - Link key refresh procedure failed.

Fix missing bt_conn_unref on encryption key refresh with error status.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

carlescufi commented:
@Vudentz @jhedberg @joerchan 

We have gone through this discussion in the past and the best solution we found then, which in my opinion continues to be the better option is the following:

1. Have a `security_changed()` connection callback with a `bt_security_t` parameter that tells the application the current security level. No `security_failed()` is needed. Instead, if the user starts a security procedure he can check the security level provided in `security_changed()` to see if the procedure is successful or not. If it is not, it will either never be called (pairing with failure) or it will be called with the same security level as before (encryption reestablishment with failure from the peer).
2. Have a `pairing_complete()`/`pairing_failed(reason)` tuple that reflects the status of a pairing procedure

This is the approach we took with the SoftDevice API and we have not had any known issues with it. This also accurately reflects the relationship between `security_changed()` and the `ENCRYPTION_CHANGE` event. At the same time it provides detailed feedback about a pairing/bonding procedure.
If the application really needs to wait on the outcome of `bt_conn_security()` it can do so by waiting on both the `security_changed()` and `pairing_*()` callbacks.
If you really think we should *always* get a `security_<>()` callback for every invocation of `bt_conn_security()` then I suggest that we stick to `security_changed()` with the same level of conn security as the one previous to the invocation of `bt_conn_security()` but I think that it is not necessary.

Vudentz commented:
> @Vudentz @jhedberg @joerchan
> 
> We have gone through this discussion in the past and the best solution we found then, which in my opinion continues to be the better option is the following:
> 
>     1. Have a `security_changed()` connection callback with a `bt_security_t` parameter that tells the application the current security level. No `security_failed()` is needed. Instead, if the user starts a security procedure he can check the security level provided in `security_changed()` to see if the procedure is successful or not. If it is not, it will either never be called (pairing with failure) or it will be called with the same security level as before (encryption reestablishment with failure from the peer).

So security_failed is not good because it is not needed, though that is the only option if you want to have different files checking if the security could not be reached or alternatively do with set_security callback though the concept is the same, we need to return to the code originating the change.

>     2. Have a `pairing_complete()`/`pairing_failed(reason)` tuple that reflects the status of a pairing procedure

These callbacks are limited to one central place, if there multiple codes attempting to set the security we would still need another set of callbacks to distribute the error.

joerchan committed:
Bluetooth: Host: Add option to force pairing in bt_conn_security

Add option to force the host to initiate pairing procedure even if the
host has encryption keys for the peer.
This option can be used to pair with a bonded peer that has deleted its
bonding information without deleting the keys. If new pairing results
in weaker keys the pairing will be aborted.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: Shell: Add force-pair option to bt security command

Add shell argument to use bt_conn_security force pairing functionality.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: SMP: Re-pairing cannot lower the security level of the bond

Make sure that a new pairing procedure with an existing bond does not
result in a security with weaker security properties.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: Shell: Add pairing failed and complete to OOB handling

Add pairing failed and complete callback to OOB authentication handlers.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: SMP: Give security changed when rejecting LTK

Give the security changed callback when the peripheral initiated
security request and peer attempted to encrypt the connection but no LTK
match was found.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@carlescufi @Vudentz @jhedberg I have updated with what we agreed upon in our call. Can you have a look?

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -413,8 +446,10 @@ struct bt_conn_cb {
 	 */
-	void (*security_changed)(struct bt_conn *conn, bt_security_t level);
+	void (*security_changed)(struct bt_conn *conn, bt_security_t level,
+				 u8_t err);
```
Using u8_t for the enum seems a bit arbitrary. I think I'd rather name the enum and use it instead. That way it also becomes clearer for the user exactly what the possible values of the parameter are. You could either just use `enum bt_security_err` but considering we have `bt_security_t`perhaps it'd be more consistent to use a typedef here, i.e. `bt_security_err_t`. If you want, you could also use `__packed` with it like we do with `bt_security_t`, which essentially makes it an `u8_t` then. That said, it's not strictly needed since I don't think we use this in any structs (where memory could be saved) and operating on a word-sized integer should be more efficient (less CPU instructions).

carlescufi replied:
I agree with this, I was going to write about it but then thought I had commented enough on cosmetic issues. I don't mind typedef or enum, or even int, but u8_t definitely feels wrong.

joerchan replied:
Yeah, makes sense, now that it is not HCI or SMP error anymore. Will switch to enum.
The typedef for bt_security_t makes sense, because it is packed. But I don't think i need packed, so will use enum instead of typedef. Seems to be more common to not typedef in zephyr.

Actually I think it is only bt_security_t and bt_addr_t, they always seems like outliers to me.

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -698,8 +733,9 @@ struct bt_conn_auth_cb {
+	 * @param reason Pairing failed reason
 	 */
-	void (*pairing_failed)(struct bt_conn *conn);
+	void (*pairing_failed)(struct bt_conn *conn, u8_t reason);
```
Same here. Using the enum name will make it clear where to look up the possible values.

joerchan replied:
switched to enum.

jhedberg commented on samples/bluetooth/peripheral_hids/src/main.c:
```diff
@@ -58,13 +58,19 @@ static void disconnected(struct bt_conn *conn, u8_t reason)
+		printk("Security changed: %s level %u", addr, level);
+	} else {
+		printk("Security failed: %s level %u reason %d", addr, level,
+		       status);
```
Why "reason" in the log string? We now have "reason", "status" and "err" for the same thing. Let's try to stick to just one if possible.

carlescufi replied:
I suggest `status` because the enum contains a `SUCCESS`

jhedberg replied:
Then the names of the enum values should also need to be changed to `BT_SECURITY_STATUS_...` which makes them rather long (well, just three characters, but the fact that we originally didn't start truncating "security" to "sec" doesn't help). I'm more or less fine either way, perhaps slightly leaning toward `err` since we do already use it elsewhere where "success" is possible. E.g. in the bt_ready and connected callbacks.

joerchan replied:
We also have BT_HCI_ERR_SUCCESS. In the connected callback we give this parameter as "err", in disconnected it is "reason" (which makes sense since it is never 0 here)
I suggest to keep this in sync with HCI error codes and use 'err' for the security_changed, and as `reason` in pairing_failed.

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -413,8 +446,10 @@ struct bt_conn_cb {
 	 *
 	 *  @param conn Connection object.
 	 *  @param level New security level of the connection.
+	 *  @param reason Security error. Zero for success, non-zero otherwise.
```
You have "reason" here, but in the definition you use "err"

joerchan replied:
Thanks. fixed.

carlescufi commented on include/bluetooth/conn.h:
```diff
@@ -413,8 +446,10 @@ struct bt_conn_cb {
 	 */
-	void (*security_changed)(struct bt_conn *conn, bt_security_t level);
+	void (*security_changed)(struct bt_conn *conn, bt_security_t level,
+				 u8_t err);
```
This should be called `status` and not `err`

joerchan replied:
Same comment as https://github.com/zephyrproject-rtos/zephyr/pull/17306#discussion_r317117617 I'd prefer to keep using `err`

jhedberg commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -107,6 +107,7 @@ struct bt_conn {
 	bt_security_t		sec_level;
 	bt_security_t		required_sec_level;
 	u8_t			encrypt;
+	u8_t                    force_pairing;
```
Why not add it to `conn->flags` instead? That way you don't need to increase the struct size.

joerchan replied:
Good idea. Switched to that instead.

joerchan commented:
@jhedberg @carlescufi Addressed (or argued) all comments.

carlescufi commented on include/bluetooth/conn.h:
```diff
@@ -413,8 +446,10 @@ struct bt_conn_cb {
 	 */
-	void (*security_changed)(struct bt_conn *conn, bt_security_t level);
+	void (*security_changed)(struct bt_conn *conn, bt_security_t level,
+				 enum bt_security_err err);
```
Please use `status`

carlescufi commented on include/bluetooth/conn.h:
```diff
@@ -323,6 +327,35 @@ int bt_conn_security(struct bt_conn *conn, bt_security_t sec);
  */
 u8_t bt_conn_enc_key_size(struct bt_conn *conn);
 
+enum bt_security_err {
```
I would rename this to `bt_security_status` and use `BT_SECURITY_STATUS_*`, because `ERR_SUCCESS` makes no sense

carlescufi commented on include/bluetooth/conn.h:
```diff
@@ -698,8 +733,10 @@ struct bt_conn_auth_cb {
 	 */
-	void (*pairing_failed)(struct bt_conn *conn);
+	void (*pairing_failed)(struct bt_conn *conn,
+			       enum bt_security_err reason);
```
s/reason/status

carlescufi commented on samples/bluetooth/peripheral_hids/src/main.c:
```diff
@@ -58,13 +58,18 @@ static void disconnected(struct bt_conn *conn, u8_t reason)
 
-static void security_changed(struct bt_conn *conn, bt_security_t level)
+static void security_changed(struct bt_conn *conn, bt_security_t level,
+			     enum bt_security_err err)
```
s/err/status

carlescufi commented on samples/bluetooth/peripheral_sc_only/src/main.c:
```diff
@@ -63,13 +63,18 @@ static void identity_resolved(struct bt_conn *conn, const bt_addr_le_t *rpa,
 
-static void security_changed(struct bt_conn *conn, bt_security_t level)
+static void security_changed(struct bt_conn *conn, bt_security_t level,
+			     enum bt_security_err err)
```
s/err/status

carlescufi commented on samples/bluetooth/peripheral_sc_only/src/main.c:
```diff
@@ -102,9 +107,9 @@ static void pairing_complete(struct bt_conn *conn, bool bonded)
 }
 
-static void pairing_failed(struct bt_conn *conn)
+static void pairing_failed(struct bt_conn *conn, enum bt_security_err reason)
```
s/reason/status

carlescufi commented on samples/boards/reel_board/mesh_badge/src/main.c:
```diff
@@ -106,9 +106,9 @@ static void pairing_complete(struct bt_conn *conn, bool bonded)
 }
 
-static void pairing_failed(struct bt_conn *conn)
+static void pairing_failed(struct bt_conn *conn, enum bt_security_err reason)
```
s/reason/status

carlescufi reviewed (changes requested):
I would argue `STATUS` is the better way of describing this. If you prefer to use `ERR` then the enum should *not* have a `SUCCESS` entry. I don't think we should use `reason`, let's be consistent

carlescufi reviewed (approved):
Approving since there has been an agreement to rename/deprecated in a later PR
